pkg/proto/stream/stream.go

```diff
@@ -0,0 +1,548 @@
+package stream
+
+import (
+	"bytes"
+	"io"
+	"math"
+
+	"github.com/golang/protobuf/proto"
+)
+
+const (
+	wtVarint          int = 0
+	wt64Bit           int = 1
+	wtLengthDelimited int = 2
+	wt32Bit           int = 5
+)
+
+// A ProtoStream supports writing protobuf data in a streaming fashion.  Its methods
+// will write their output to the wrapped `io.Writer`.  Zero values are not included.
+//
+// ProtoStream instances are *not* threadsafe and *not* re-entrant.
+type ProtoStream struct {
```

Comment:
An interface would introduce a dynamic dispatch (indirect jump).  Since this is used in hot loops, that is potentially a high cost.  Types with no public [are](https://pkg.go.dev/bytes#Buffer) [pretty](https://pkg.go.dev/math/rand#Rand) [common](https://pkg.go.dev/os#File).

Reply:
Yeah fair points